app/models/targeting.rb

```diff
@@ -0,0 +1,28 @@
+class Targeting < ActiveRecord::Base
+
+  belongs_to :user
+  belongs_to :bookmark
+
+  has_many :targeting_hosts, :dependent => :destroy
+  has_many :hosts, :through => :targeting_hosts
+  has_many :template_invocations, :dependent => :destroy
+
+  validates :bookmark_id, :presence => {:message => N_("must be specified if search query is blank")},
+            :if => Proc.new { |a| a.search_query.blank? }
+
+  validate :bookmark_or_query
+
+
+  def resolve_hosts!
+      query = self.bookmark ? self.bookmark.query : self.search_query
```

Comment:
This may be my personal style, but I prefer using self personally as its more explicit.  Using .try() is more elegant in this case, will update